logics/impls/controller/mod.rs

```diff
@@ -718,24 +694,232 @@ impl<T: Storage<Data>> Internal for T {
         pool: AccountId,
         redeemer: AccountId,
         redeem_amount: Balance,
-        pool_attribute: Option<PoolAttributes>,
+        pool_attributes: Option<PoolAttributesForWithdrawValidation>,
     ) -> Result<()> {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
-        } else {
-            None
+        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
```

Comment:
I understood that this function is written in the following flow.

```rust
    default fn _redeem_allowed(
        ...
    ) -> Result<()> {
        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
        let (shortfall, account_data, asset_price, liquidation_threshold) = {
            ...

            //// [1] Decomposition of pool_attribute
            if let Some(pool_attribute) = pool_attributes {
                ...
            }

            //// [2] Loop through all tokens on the market...
            //// Collect inputs for the following calculations
            ////// capacity calculation (※1)
            ////// health_factor calculation with liquidation_threshold (※2)
            for asset in markets {
                ...
            }

            //// [3] capacity calculation (※1)
            let (sum_collateral, sum_borrow_plus_effect) =
                get_hypothetical_account_liquidity(GetHypotheticalAccountLiquidityInput {
                    asset_params,
                    token_modify: Some(pool),
                    redeem_tokens: redeem_amount,
                    borrow_amount: 0,
                });

            ...

            //// [4] preliminary calculation/data collection as input for health_factor calculation with liquidation_threshold
            let health_factor = calculate_health_factor_from_balances(
                total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_liquidation_threshold,
            );
            let account_data = AccountData {
                total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_ltv,
                avg_liquidation_threshold,
                health_factor,
            };
            ...

            value
        };

        ...

        //// [5] health_factor calculation with liquidation_threshold (※2)
        let balance_decrease_allowed_result =
            balance_decrease_allowed(BalanceDecreaseAllowedParam {
                total_collateral_in_base_currency: account_data.total_collateral_in_base_currency,
                total_debt_in_base_currency,
                avg_liquidation_threshold: account_data.avg_liquidation_threshold,
                amount_in_base_currency_unit: U256::from(redeem_amount),
                asset_price: U256::from(asset_price),
                liquidation_threshold: U256::from(liquidation_threshold),
            });

        ...

        Ok(())
    }
```

Due to the large number of lines of code in this one function (over 200 lines), we have the following issues
- There are variables that appear only a little but are valid in scope for a long time
- There is a lot of code contained in one nest (important processes [1][2][3][4] are in one nest)

I would like to improve this in one of the following ways

**Option 1: organize nesting & order variable declarations & return values**

Separate the nests containing [1][2][3][4].
And the processes for capacity calculation and health_factor calculation with liquidation_threshold.

Here's the image.

```rust
    default fn _redeem_allowed(...) -> Result<()> {
        let oracle = self._oracle().ok_or(Error::OracleIsNotSet)?;
        let caller = Self::env().caller();

        let mut total_collateral_in_base_currency = U256::from(0);
        let mut avg_ltv = U256::from(0);
        let mut avg_liquidation_threshold = U256::from(0);
        let mut total_debt_in_base_currency = U256::from(0);

        //// [1] Decomposition of pool_attribute
        let (asset_price, liquidation_threshold, skip_pool) =
            if let Some(pool_attribute) = pool_attributes {
                // Collect input for capacity calculation
                asset_params.push(HypotheticalAccountLiquidityCalculationParam { ... });

                // Collect input for health_factor calculation with liquidation_threshold
                let liquidation_threshold = pool_attribute.liquidation_threshold;

                let compounded_liquidity_balance = pool_attribute.account_balance;
                if compounded_liquidity_balance != 0 { ... }

                let borrow_balance_stored = pool_attribute.account_borrow_balance;
                if borrow_balance_stored != 0 { ... }

                (oracle_price, liquidation_threshold, pool)
            } else {
                (0, 0, caller)
            };
        
        //// [2] Loop through all tokens on the market...
        for asset in markets {
          ...

          let (compounded_liquidity_balance, borrow_balance_stored, exchange_rate_mantissa) =
              PoolRef::get_account_snapshot(&asset, redeemer);
          //// NOTE: do not use account_asset function as it is making duplicated cross-contract calling leads to high gas.
          if compounded_liquidity_balance == 0 && borrow_balance_stored == 0 {
              continue
          }

          // Collect input for capacity calculation
          let decimals = ...;
          let oracle_price: u128 = ...;
          let collateral_factor_mantissa = ...;

          ...
          asset_params.push(HypotheticalAccountLiquidityCalculationParam { ... });

          // Collect input for health_factor calculation with liquidation_threshold
          let ltv = U256::from(collateral_factor_mantissa);
          let liquidation_threshold = PoolRef::liquidation_threshold(&asset);

          if compounded_liquidity_balance != 0 {
              let liquidity_balance_eth = ...;
              total_collateral_in_base_currency = ...;
              avg_ltv = ...;
              avg_liquidation_threshold = ...;
          }

          if borrow_balance_stored != 0 {
              let borrow_balance_eth = ...
              total_debt_in_base_currency = ...;
          }
      }

      //// [3] capacity calculation (※1)
      let (sum_collateral, sum_borrow_plus_effect) =
          get_hypothetical_account_liquidity(GetHypotheticalAccountLiquidityInput { ... });
      let shortfall = if sum_collateral > sum_borrow_plus_effect {
          U256::from(0)
      } else {
          sum_borrow_plus_effect.sub(sum_collateral)
      };
      if !shortfall.is_zero() {
          return Err(Error::InsufficientLiquidity)
      }

      //// [4] preliminary calculation/data collection as input for health_factor calculation with liquidation_threshold
      avg_ltv = if total_collateral_in_base_currency.is_zero() {
          U256::from(0)
      } else {
          avg_ltv.div(total_collateral_in_base_currency)
      };
      avg_liquidation_threshold = if total_collateral_in_base_currency.is_zero() {
          U256::from(0)
      } else {
          avg_liquidation_threshold.div(total_collateral_in_base_currency)
      };
      let health_factor = calculate_health_factor_from_balances( ... );
      let account_data = AccountData { ... };

      //// [5] health_factor calculation with liquidation_threshold (※2)
      let total_debt_in_base_currency = account_data.total_debt_in_base_currency;
      if total_debt_in_base_currency.is_zero() {
          return Ok(())
      }
      let balance_decrease_allowed_result =
          balance_decrease_allowed(BalanceDecreaseAllowedParam { ... });
      if !balance_decrease_allowed_result {
          return Err(Error::BalanceDecreaseNotAllowed)
      }

      ...
    }
```

Nesting involving [1][2][3][4] as shown above is gone,
I think it will reduce the number of updatable variables (except for arg_xxx at the beginning) and make it easier to see where the variables are for.

**Option 2: Write a test that cuts the process into a function and shares it with existing logic**

Option1 may be better for code brevity, but may be preferable for functional quality.

Most of the code here is
- Check for surplus capacity (flow up to call _get_hypothetical_account_liquidity in Controller#_borrow_allowed)
- Controller#__balance_decrease_allowed
and reordered to reduce the number of Pool calls, etc.

Therefore, we should check the quality of each location function (ex: redeem, transfer...) at the same time by creating functions that share code with these existing logics and testing them.

Reply:
Do you want me to separate the contents into functions?

Reply:
it would be better if the code could be separated into functions (Option 2) for testability, but since the biggest problem is low readability due to nest and variable scopes.
It would be fine to maintain this function in the manner of Option 1, up to and including rearrangement of the description order and scope.